File: app/src/main/java/com/example/teluguclockwidget/TeluguClockWidgetConfigureActivity.java
package com.example.teluguclockwidget;

import android.app.Activity;
import android.appwidget.AppWidgetManager;
import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.graphics.Color;
import android.os.Bundle;
import android.view.Gravity;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;
import android.widget.RadioButton;
import android.widget.RadioGroup;
import android.widget.CheckBox;
import android.util.Log;

public class TeluguClockWidgetConfigureActivity extends Activity {

    private int mAppWidgetId = AppWidgetManager.INVALID_APPWIDGET_ID;
    private EditText mFontSizeEditText;
    private EditText mFontColorEditText;
    private RadioGroup mHorizontalGravityRadioGroup;
    private RadioGroup mVerticalGravityRadioGroup;
    private RadioGroup mTimeFormatRadioGroup;
    private CheckBox mShowAmPmIconCheckBox;
    private CheckBox mFontBoldCheckBox;
    private CheckBox mShowBackgroundOverlayCheckBox;
    private EditText mBackgroundOverlayColorEditText;
    private CheckBox mShowDateCheckBox;
    private CheckBox mShowTextShadowCheckBox;


    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_telugu_clock_widget_configure);

        // Set the result to CANCELED.  This will cause the widget host to cancel
        // out of the widget placement if the user presses the back button.
        setResult(RESULT_CANCELED);

        mFontSizeEditText = findViewById(R.id.font_size_edit_text);
        mFontColorEditText = findViewById(R.id.font_color_edit_text);
        mHorizontalGravityRadioGroup = findViewById(R.id.horizontal_gravity_radio_group);
        mVerticalGravityRadioGroup = findViewById(R.id.vertical_gravity_radio_group);
        mTimeFormatRadioGroup = findViewById(R.id.time_format_radio_group);
        mShowAmPmIconCheckBox = findViewById(R.id.show_am_pm_icon_checkbox);
        mFontBoldCheckBox = findViewById(R.id.font_bold_checkbox);
        mShowBackgroundOverlayCheckBox = findViewById(R.id.show_background_overlay_checkbox);
        mBackgroundOverlayColorEditText = findViewById(R.id.background_overlay_color_edit_text);
        mShowDateCheckBox = findViewById(R.id.show_date_checkbox);
        mShowTextShadowCheckBox = findViewById(R.id.show_text_shadow_checkbox);


        Button saveButton = findViewById(R.id.save_button);
        saveButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                final Context context = TeluguClockWidgetConfigureActivity.this;

                // When the button is clicked, save the string in our prefs and return that they
                // clicked OK.
                String fontSizeStr = mFontSizeEditText.getText().toString();
                String fontColorStr = mFontColorEditText.getText().toString();

                int fontSize = 48;
                if (!fontSizeStr.isEmpty()) {
                    fontSize = Integer.parseInt(fontSizeStr);
                }

                int fontColor = Color.WHITE;
                if (!fontColorStr.isEmpty()) {
                    try {
                        fontColor = Color.parseColor(fontColorStr);
                    } catch (IllegalArgumentException e) {
                        // Invalid color format, use default
                    }
                }

                int selectedHorizontalGravityId = mHorizontalGravityRadioGroup.getCheckedRadioButtonId();
                int horizontalGravity = Gravity.CENTER_HORIZONTAL;
                if (selectedHorizontalGravityId == R.id.left_radio_button) {
                    horizontalGravity = Gravity.LEFT;
                } else if (selectedHorizontalGravityId == R.id.right_radio_button) {
                    horizontalGravity = Gravity.RIGHT;
                }

                int selectedVerticalGravityId = mVerticalGravityRadioGroup.getCheckedRadioButtonId();
                int verticalGravity = Gravity.CENTER_VERTICAL;
                if (selectedVerticalGravityId == R.id.top_radio_button) {
                    verticalGravity = Gravity.TOP;
                } else if (selectedVerticalGravityId == R.id.bottom_radio_button) {
                    verticalGravity = Gravity.BOTTOM;
                }

                int selectedTimeFormatId = mTimeFormatRadioGroup.getCheckedRadioButtonId();
                boolean is12HourFormat = (selectedTimeFormatId == R.id.format_12_hour_radio_button);
                boolean showAmPmIcon = mShowAmPmIconCheckBox.isChecked();
                boolean isFontBold = mFontBoldCheckBox.isChecked();
                boolean showBackgroundOverlay = mShowBackgroundOverlayCheckBox.isChecked();
                String backgroundOverlayColorStr = mBackgroundOverlayColorEditText.getText().toString();
                int backgroundOverlayColor = Color.TRANSPARENT;
                if (!backgroundOverlayColorStr.isEmpty()) {
                    try {
                        backgroundOverlayColor = Color.parseColor(backgroundOverlayColorStr);
                    } catch (IllegalArgumentException e) {
                        // Invalid color format, use default
                    }
                }
                boolean showDate = mShowDateCheckBox.isChecked();
                boolean showTextShadow = mShowTextShadowCheckBox.isChecked();

                final Bundle options = new Bundle();
                options.putInt("font_size", fontSize);
                options.putInt("font_color", fontColor);
                options.putInt("horizontal_gravity", horizontalGravity);
                options.putInt("vertical_gravity", verticalGravity);
                options.putBoolean("is_12_hour_format", is12HourFormat);
                options.putBoolean("show_am_pm_icon", showAmPmIcon);
                options.putBoolean("is_font_bold", isFontBold);
                options.putBoolean("show_background_overlay", showBackgroundOverlay);
                options.putInt("background_overlay_color", backgroundOverlayColor);
                options.putBoolean("show_date", showDate);
                options.putBoolean("show_text_shadow", showTextShadow);

                savePrefs(context, mAppWidgetId, fontSize, fontColor, horizontalGravity, verticalGravity, is12HourFormat, showAmPmIcon, isFontBold, showBackgroundOverlay, backgroundOverlayColor, showDate, showTextShadow);

                // It is the responsibility of the configuration activity to update the app widget
                AppWidgetManager appWidgetManager = AppWidgetManager.getInstance(context);
                TeluguClockWidget.updateAppWidgetWithPrefs(context, appWidgetManager, mAppWidgetId, options);

                // Make sure we pass back the original appWidgetId
                Intent resultValue = new Intent();
                resultValue.putExtra(AppWidgetManager.EXTRA_APPWIDGET_ID, mAppWidgetId);
                setResult(RESULT_OK, resultValue);
                finish();
            }
        });

        // Find the widget id from the intent.
        Intent intent = getIntent();
        Bundle extras = intent.getExtras();
        if (extras != null) {
            mAppWidgetId = extras.getInt(
                    AppWidgetManager.EXTRA_APPWIDGET_ID, AppWidgetManager.INVALID_APPWIDGET_ID);
        }

        // If this activity was started with an intent without an app widget ID, finish with an error.
        if (mAppWidgetId == AppWidgetManager.INVALID_APPWIDGET_ID) {
            finish();
            return;
        }

        loadPrefs(this, mAppWidgetId);
    }

    private void loadPrefs(Context context, int appWidgetId) {
        SharedPreferences prefs = context.getSharedPreferences("widget_prefs_" + appWidgetId, Context.MODE_PRIVATE);
        int fontSize = prefs.getInt("font_size", 48);
        int fontColor = prefs.getInt("font_color", Color.WHITE);
        int horizontalGravity = prefs.getInt("horizontal_gravity", Gravity.CENTER_HORIZONTAL);
        int verticalGravity = prefs.getInt("vertical_gravity", Gravity.CENTER_VERTICAL);
        boolean is12HourFormat = prefs.getBoolean("is_12_hour_format", true); // Default to 12-hour format
        boolean showAmPmIcon = prefs.getBoolean("show_am_pm_icon", false); // Default to not showing AM/PM icon
        boolean isFontBold = prefs.getBoolean("is_font_bold", false); // Default to not bold
        boolean showBackgroundOverlay = prefs.getBoolean("show_background_overlay", false);
        int backgroundOverlayColor = prefs.getInt("background_overlay_color", Color.TRANSPARENT);
        boolean showDate = prefs.getBoolean("show_date", false);
        boolean showTextShadow = prefs.getBoolean("show_text_shadow", false);

        mFontSizeEditText.setText(String.valueOf(fontSize));
        mFontColorEditText.setText(String.format("#%06X", (0xFFFFFF & fontColor)));

        if (horizontalGravity == Gravity.LEFT) {
            mHorizontalGravityRadioGroup.check(R.id.left_radio_button);
        } else if (horizontalGravity == Gravity.RIGHT) {
            mHorizontalGravityRadioGroup.check(R.id.right_radio_button);
        } else {
            mHorizontalGravityRadioGroup.check(R.id.center_horizontal_radio_button);
        }

        if (verticalGravity == Gravity.TOP) {
            mVerticalGravityRadioGroup.check(R.id.top_radio_button);
        } else if (verticalGravity == Gravity.BOTTOM) {
            mVerticalGravityRadioGroup.check(R.id.bottom_radio_button);
        } else {
            mVerticalGravityRadioGroup.check(R.id.center_vertical_radio_button);
        }

        if (is12HourFormat) {
            mTimeFormatRadioGroup.check(R.id.format_12_hour_radio_button);
        } else {
            mTimeFormatRadioGroup.check(R.id.format_24_hour_radio_button);
        }
        mShowAmPmIconCheckBox.setChecked(showAmPmIcon);
        mFontBoldCheckBox.setChecked(isFontBold);
        mShowBackgroundOverlayCheckBox.setChecked(showBackgroundOverlay);
        mBackgroundOverlayColorEditText.setText(String.format("#%08X", backgroundOverlayColor));
        mShowDateCheckBox.setChecked(showDate);
        mShowTextShadowCheckBox.setChecked(showTextShadow);

        Log.d("WidgetConfig", "Loaded prefs for widget " + appWidgetId + ": " +
                "fontSize=" + fontSize +
                ", fontColor=" + String.format("#%08X", fontColor) +
                ", horizontalGravity=" + horizontalGravity +
                ", verticalGravity=" + verticalGravity +
                ", is12HourFormat=" + is12HourFormat +
                ", showAmPmIcon=" + showAmPmIcon +
                ", isFontBold=" + isFontBold +
                ", showBackgroundOverlay=" + showBackgroundOverlay +
                ", backgroundOverlayColor=" + String.format("#%08X", backgroundOverlayColor) +
                ", showDate=" + showDate +
                ", showTextShadow=" + showTextShadow);
    }

    private void savePrefs(Context context, int appWidgetId, int fontSize, int fontColor, int horizontalGravity, int verticalGravity, boolean is12HourFormat, boolean showAmPmIcon, boolean isFontBold, boolean showBackgroundOverlay, int backgroundOverlayColor, boolean showDate, boolean showTextShadow) {
        SharedPreferences.Editor prefs = context.getSharedPreferences("widget_prefs_" + appWidgetId, Context.MODE_PRIVATE).edit();
        prefs.putInt("font_size", fontSize);
        prefs.putInt("font_color", fontColor);
        prefs.putInt("horizontal_gravity", horizontalGravity);
        prefs.putInt("vertical_gravity", verticalGravity);
        prefs.putBoolean("is_12_hour_format", is12HourFormat);
        prefs.putBoolean("show_am_pm_icon", showAmPmIcon);
        prefs.putBoolean("is_font_bold", isFontBold);
        prefs.putBoolean("show_background_overlay", showBackgroundOverlay);
        prefs.putInt("background_overlay_color", backgroundOverlayColor);
        prefs.putBoolean("show_date", showDate);
        prefs.putBoolean("show_text_shadow", showTextShadow);
        prefs.apply();

        Log.d("WidgetConfig", "Saved prefs for widget " + appWidgetId + ": " +
                "fontSize=" + fontSize +
                ", fontColor=" + String.format("#%08X", fontColor) +
                ", horizontalGravity=" + horizontalGravity +
                ", verticalGravity=" + verticalGravity +
                ", is12HourFormat=" + is12HourFormat +
                ", showAmPmIcon=" + showAmPmIcon +
                ", isFontBold=" + isFontBold +
                ", showBackgroundOverlay=" + showBackgroundOverlay +
                ", backgroundOverlayColor=" + String.format("#%08X", backgroundOverlayColor) +
                ", showDate=" + showDate +
                ", showTextShadow=" + showTextShadow);
    }
}

File: app/src/main/java/com/example/teluguclockwidget/TeluguClockWidgetConfigureActivity.kt
package com.example.teluguclockwidget

import android.app.Activity
import android.app.AlertDialog
import android.appwidget.AppWidgetManager
import android.content.Intent
import android.graphics.Color
import android.graphics.drawable.ColorDrawable
import android.os.Bundle
import android.text.Editable
import android.text.TextWatcher
import android.view.View
import android.widget.*
import androidx.appcompat.app.AppCompatActivity
import androidx.core.content.ContextCompat
import java.util.Calendar

class TeluguClockWidgetConfigureActivity : AppCompatActivity() {

    private var appWidgetId = AppWidgetManager.INVALID_APPWIDGET_ID
    
    private lateinit var fontSizeEt: EditText
    private lateinit var colorPickerBtn: Button // Changed from colorPickerButton
    private lateinit var colorPreview: View
    private lateinit var fontSpinner: Spinner
    private lateinit var boldCb: CheckBox
    private lateinit var shadowCb: CheckBox
    private lateinit var overlayCb: CheckBox
    private lateinit var showDateCb: CheckBox
    private lateinit var dateFormatRg: RadioGroup
    private lateinit var overlayThemeRg: RadioGroup
    private lateinit var timeFormatRg: RadioGroup
    private lateinit var previewImage: ImageView
    private lateinit var fonts: Array<String>

    private var selectedFontColor: Int = Color.WHITE

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_telugu_clock_widget_configure)

        setResult(Activity.RESULT_CANCELED)

        // Initialize views
        fontSizeEt = findViewById(R.id.font_size_edit_text)
        colorPickerBtn = findViewById(R.id.color_picker_button) // Changed from colorPickerButton
        colorPreview = findViewById(R.id.color_preview)
        fontSpinner = findViewById(R.id.font_family_spinner)
        boldCb = findViewById(R.id.font_bold_checkbox)
        shadowCb = findViewById(R.id.show_text_shadow_checkbox)
        overlayCb = findViewById(R.id.show_background_overlay_checkbox)
        showDateCb = findViewById(R.id.show_date_checkbox)
        dateFormatRg = findViewById(R.id.date_format_radio_group)
        overlayThemeRg = findViewById(R.id.overlay_theme_radio_group)
        timeFormatRg = findViewById(R.id.time_format_radio_group)
        val saveBtn = findViewById<Button>(R.id.save_button)
        previewImage = findViewById(R.id.preview_image)

        fonts = resources.getStringArray(R.array.font_family_array)
        fontSpinner.adapter = ArrayAdapter(this, android.R.layout.simple_spinner_item, fonts).apply {
            setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item)
        }

        // get widget id
        intent.extras?.let {
            appWidgetId = it.getInt(AppWidgetManager.EXTRA_APPWIDGET_ID, AppWidgetManager.INVALID_APPWIDGET_ID)
        }
        if (appWidgetId == AppWidgetManager.INVALID_APPWIDGET_ID) {
            finish()
            return
        }

        // load prefs
        val prefs = WidgetPrefs(this, appWidgetId)
        fontSizeEt.setText(prefs.fontSize.toString())
        selectedFontColor = prefs.fontColor
        colorPreview.setBackgroundColor(selectedFontColor)
        boldCb.isChecked = prefs.bold
        shadowCb.isChecked = prefs.shadow
        overlayCb.isChecked = prefs.overlay
        showDateCb.isChecked = prefs.showDate
        timeFormatRg.check(if (prefs.is12Hour) R.id.format_12_hour_radio_button else R.id.format_24_hour_radio_button)
        dateFormatRg.check(when (prefs.dateFormat) {
            1 -> R.id.date_format_1
            else -> R.id.date_format_2
        })
        overlayThemeRg.check(when (prefs.overlayTheme) {
            "black" -> R.id.overlay_black
            else -> R.id.overlay_white
        })

        // set spinner selection
        val idx = fonts.indexOf(prefs.fontFamily)
        if (idx >= 0) fontSpinner.setSelection(idx)

        // Setup live preview listeners
        setupLivePreview()
        
        // Initial preview
        updatePreview()

        saveBtn.setOnClickListener {
            saveSettings()
        }

        // Color picker button listener
        colorPickerBtn.setOnClickListener { // Changed from colorPickerButton
            showColorPickerDialog("Select Font Color", selectedFontColor) { color ->
                selectedFontColor = color
                colorPreview.setBackgroundColor(selectedFontColor)
                updatePreview()
            }
        }
    }

    private fun showColorPickerDialog(title: String, initialColor: Int, onColorSelected: (Int) -> Unit) {
        val colors = arrayOf(
            Color.BLACK, Color.DKGRAY, Color.GRAY, Color.LTGRAY, Color.WHITE,
            Color.RED, Color.GREEN, Color.BLUE, Color.YELLOW, Color.CYAN, Color.MAGENTA,
            ContextCompat.getColor(this, android.R.color.holo_red_light),
            ContextCompat.getColor(this, android.R.color.holo_green_light),
            ContextCompat.getColor(this, android.R.color.holo_blue_light),
            ContextCompat.getColor(this, android.R.color.holo_orange_light),
            ContextCompat.getColor(this, android.R.color.holo_purple)
        )
        val colorNames = arrayOf(
            "Black", "Dark Gray", "Gray", "Light Gray", "White",
            "Red", "Green", "Blue", "Yellow", "Cyan", "Magenta",
            "Light Red", "Light Green", "Light Blue", "Light Orange", "Light Purple"
        )

        val builder = AlertDialog.Builder(this)
        builder.setTitle(title)

        val layout = LinearLayout(this).apply {
            orientation = LinearLayout.VERTICAL
            setPadding(50, 30, 50, 30)
        }

        val radioGroup = RadioGroup(this)
        radioGroup.orientation = LinearLayout.VERTICAL

        colors.forEachIndexed { index, color ->
            val radioButton = RadioButton(this).apply {
                text = colorNames[index]
                id = index
                if (color == initialColor) {
                    isChecked = true
                }
            }
            radioGroup.addView(radioButton)
        }
        layout.addView(radioGroup)
        builder.setView(layout)

        builder.setPositiveButton("Select") { dialog, _ ->
            val selectedId = radioGroup.checkedRadioButtonId
            if (selectedId != -1) {
                onColorSelected(colors[selectedId])
            }
            dialog.dismiss()
        }
        builder.setNegativeButton("Cancel") { dialog, _ ->
            dialog.cancel()
        }
        builder.show()
    }

    private fun setupLivePreview() {
        // Font size changes
        fontSizeEt.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(s: CharSequence?, start: Int, count: Int, after: Int) {}
            override fun onTextChanged(s: CharSequence?, start: Int, before: Int, count: Int) {}
            override fun afterTextChanged(s: Editable?) {
                updatePreview()
            }
        })

        // Font family changes
        fontSpinner.onItemSelectedListener = object : AdapterView.OnItemSelectedListener {
            override fun onItemSelected(parent: AdapterView<*>?, view: View?, position: Int, id: Long) {
                updatePreview()
            }
            override fun onNothingSelected(parent: AdapterView<*>?) {}
        }

        // Checkbox changes
        val checkboxListener = CompoundButton.OnCheckedChangeListener { _, _ ->
            updatePreview()
        }
        boldCb.setOnCheckedChangeListener(checkboxListener)
        shadowCb.setOnCheckedChangeListener(checkboxListener)
        overlayCb.setOnCheckedChangeListener(checkboxListener)
        showDateCb.setOnCheckedChangeListener(checkboxListener)

        // Time format changes
        timeFormatRg.setOnCheckedChangeListener { _, _ ->
            updatePreview()
        }

        // Date format changes
        dateFormatRg.setOnCheckedChangeListener { _, _ ->
            updatePreview()
        }

        // Overlay theme changes
        overlayThemeRg.setOnCheckedChangeListener { _, _ ->
            updatePreview()
        }
    }

    private fun updatePreview() {
        try {
            // Get current settings
            val prefs = WidgetPrefs(this, appWidgetId)
            val fontSize = fontSizeEt.text.toString().toIntOrNull() ?: prefs.fontSize
            val fontColor = selectedFontColor // Use selectedFontColor
            val fontFamily = fonts[fontSpinner.selectedItemPosition]
            val bold = boldCb.isChecked
            val shadow = shadowCb.isChecked
            val showDate = showDateCb.isChecked
            val dateFormat = when (dateFormatRg.checkedRadioButtonId) { // Get dateFormat
                R.id.date_format_1 -> 1
                else -> 2
            }
            val is12h = timeFormatRg.checkedRadioButtonId == R.id.format_12_hour_radio_button

            // Generate preview time
            val cal = Calendar.getInstance()
            var hour = if (is12h) cal.get(Calendar.HOUR) else cal.get(Calendar.HOUR_OF_DAY)
            if (is12h && hour == 0) hour = 12
            val minute = cal.get(Calendar.MINUTE)

            val formatted = String.format("%02d:%02d", hour, minute)
            val telugu = ClockBitmapRenderer.toTelugu(formatted)

            // Render preview bitmap
            val bmp = ClockBitmapRenderer.render(
                this,
                telugu,
                fontSize,
                fontColor,
                bold,
                shadow,
                fontFamily,
                showDate,
                dateFormat // Pass dateFormat
            )

            // Set preview image
            previewImage.setImageBitmap(bmp)
            
            // Remove preview background logic, as overlay is handled by TeluguClockWidget.kt
            previewImage.setBackgroundColor(Color.TRANSPARENT)
            previewImage.setPadding(0, 0, 0, 0)

        } catch (e: Exception) {
            // If preview fails, show error message
            Toast.makeText(this, "Preview error: ${e.message}", Toast.LENGTH_SHORT).show()
        }
    }

    private fun saveSettings() {
        val prefs = WidgetPrefs(this, appWidgetId)
        
        val fontSize = fontSizeEt.text.toString().toIntOrNull() ?: prefs.fontSize
        val fontColor = selectedFontColor // Use selectedFontColor
        val fontFamily = fonts[fontSpinner.selectedItemPosition]
        val bold = boldCb.isChecked
        val shadow = shadowCb.isChecked
        val overlay = overlayCb.isChecked
        val showDate = showDateCb.isChecked
        val dateFormat = when (dateFormatRg.checkedRadioButtonId) { // Get dateFormat
            R.id.date_format_1 -> 1
            else -> 2
        }
        val overlayTheme = when (overlayThemeRg.checkedRadioButtonId) { // Get overlayTheme
            R.id.overlay_black -> "black"
            else -> "white"
        }
        val is12h = timeFormatRg.checkedRadioButtonId == R.id.format_12_hour_radio_button

        // save
        prefs.fontSize = fontSize
        prefs.fontColor = fontColor // Save selectedFontColor
        prefs.fontFamily = fontFamily
        prefs.bold = bold
        prefs.shadow = shadow
        prefs.overlay = overlay
        prefs.overlayTheme = overlayTheme // Save overlayTheme
        prefs.is12Hour = is12h
        prefs.showDate = showDate
        prefs.dateFormat = dateFormat // Save dateFormat

        // update widget immediately
        val mgr = AppWidgetManager.getInstance(this)
        TeluguClockWidget.updateWidget(this, mgr, appWidgetId)

        // return result
        val result = Intent().apply { 
            putExtra(AppWidgetManager.EXTRA_APPWIDGET_ID, appWidgetId) 
        }
        setResult(Activity.RESULT_OK, result)
        finish()
    }
}